### PR TITLE
Added Lambda "Quantization Level" Parameter

### PR DESCRIPTION
**Pull request by mmoffatt2**
Adds a quantization level parameter to allow for more gradual introduction of quantization rather than applying it abruptly to the weights and activations for each tensor. Quantization Level = 0 means that there is no quantization happening. Quantization Level = 1 means full quantization. This quantization level is only applied during training to the activations and linears.

### CODE REVIEW

File: variations/linear_variations.py [hunk 1]
```diff
@@ -47,7 +50,7 @@ def training_quantized_forward(self, input):
         assert self.training, "Should be called only during training"
 
         # Applies the fake quantization to the weights
-        self._fake_quantized_weight = _fake_quantize(self.weight, self.weight_bits, self.quant_method)
+        self._fake_quantized_weight = _fake_quantize(self.weight, self, self.weight_bits, self.quant_method)
```
Comment (gkielian) [inline variations/linear_variations.py:53]: Why is self added as an input to _fake_quantize?

File: quantization/quantize.py [hunk 2]
```diff
@@ -137,14 +152,17 @@ def forward(ctx, input, bits=7, quantization_method="affine_quant"):
...
+            print("quant level: ", quant_level)
+        return input + quant_level * (dequantize(zero_point, norm, quantized_weight) - input).detach()
 
     @staticmethod
     def backward(ctx, grad_output):
         # Straight-Through Estimator (STE): passes grad_output through as the gradient with respect to the input
         # gradient is approximated by simply passing the gradient from the output directly to the input, 
         # ignoring the quantization operation
-        return grad_output, None, None
+        return grad_output, None, None, None
```
Comment (gkielian) [inline quantization/quantize.py:165]: Why is another None added to this backward pass?

File: quantization/quantize.py [hunk 3]
```diff
@@ -124,7 +139,7 @@ class FakeLinearQuantizationFunction(torch.autograd.Function):
     """
 
     @staticmethod
-    def forward(ctx, input, bits=7, quantization_method="affine_quant"):
+    def forward(ctx, input, obj, bits=7, quantization_method="affine_quant"):
```
Comment (gkielian) [inline quantization/quantize.py:142]: I see so the object is here due to being the last none named value.

Something feels off about this...

...After reflection, we should really either just pass "self" to the function and get all of the fields from the "self.bits" etc _or_ pass the needed values from the object.

I think for now let's do the latter, though let's investigate whether this is best practices in python (not to pass whole large object).

Please check on this.

Then, if passing large objects to functions is indeed not best practice, pass each of the fields you'll need individually.

Lastly add a check to only perform the quant_level modificaitons if quant_level is active (maybe set default quant_level to None in gpt_conf.py and train.py, and detect here is `if quant_level is not None`)

File: quantization/quantize.py [hunk 4]
```diff
@@ -8,6 +8,14 @@ def set_dtype(bits):
     else:
         return torch.int8
     
+def calculate_quant_level(obj, iter_num):
+    if not obj.training:
+        return 1
+    if obj.quant_level.isnumeric():
```
Comment (gkielian) [inline quantization/quantize.py:14]: We should have some method of setting a  quantization scheduler or something to ease into full quantization.

That that should gate whether quant_level is checked and updated.

File: model.py [hunk 5]
```diff
@@ -765,9 +779,9 @@ def forward(self, idx, targets=None):
         for block in self.transformer.h:
             # Propagate tokens through layers
             if self.config.use_gradient_checkpointing:
-                x = checkpoint.checkpoint(block, x, use_reentrant=self.config.recompute_backward_pass)
+                x = checkpoint.checkpoint(block, x, use_reentrant=self.config.recompute_backward_pass, **kwargs)
```
Comment (gkielian) [inline model.py:782]: Do we have to put **kwargs here?

File: model.py [hunk 6]
```diff
@@ -738,7 +752,7 @@ def export_scale_matrices(self, file_path):
         np.savez(file_path, scale_up=scale_up_matrix, scale_down=scale_down_matrix)
         print(f"Scale matrices saved to {file_path}")
 
-    def forward(self, idx, targets=None):
+    def forward(self, idx, targets=None, **kwargs):
```
Comment (gkielian) [inline model.py:755]: Do we have to put **kwargs?

File: model.py [hunk 7]
```diff
@@ -511,22 +525,22 @@ def __init__(self, config, mlp=None, attn=None):
         else:
             self.mlp = mlp
 
-    def forward(self, x):
```
Comment (gkielian) [inline model.py:514]: Check if we have to put **kwargs for what you're going for, I don't think we'll need to, since keyword arguments if not named, should be set to the default (e.g. `set_quantizatin(bits=8)` can be called with `set_quantization`) in the function definition.

File: model.py [hunk 8]
```diff
@@ -432,11 +444,13 @@ def __init__(self, config):
 
         self.dropout = nn.Dropout(config.dropout)
 
-    def forward(self, x):
+    def forward(self, x, **kwargs):
```
Comment (gkielian) [inline model.py:447]: What is **kwargs used for here?

File: model.py [hunk 9]
```diff
@@ -258,13 +263,15 @@ def __init__(self, config, fire_pos_enc=None):
                                         .view(1, 1, config.block_size, config.block_size))
 
 
-    def forward(self, x):
+    def forward(self, x, **kwargs):
```
Comment (gkielian) [inline model.py:266]: Why is **kwargs added here?

File: gpt_conf.py [hunk 10]
```diff
@@ -3,6 +3,8 @@
 import json
 @dataclass
 class GPTConfig:
+    max_iters: int = 3500
+    eval_interval: int = 250
```
Comment (gkielian) [inline gpt_conf.py:7]: Let's leave these as default.

File: gpt_conf.py [hunk 11]
```diff
@@ -203,6 +205,7 @@ class GPTConfig:
     linear_std_init: float= 0.02
 
     # Quantizations
+    quant_level: str = 1
```
Comment (gkielian) [inline gpt_conf.py:208]: Why is quant_level a string?

**Review by gkielian — COMMENTED** [review]
After reading through, have feedback and questions.

**Comment by gkielian** [comment]
From discussion change kwargs to iter_num directly.

And also provide a demo script that shows difference with and without gradual quantization with tinystories (verifying inference too)

File: quantization/quantize.py [hunk 12]
```diff
@@ -102,20 +110,27 @@ def dequantize(zero_point, scale, tensor, causal_mask=False):
...
+    if causal_mask:
+        # Create a mask for the upper triangular part
+        upper_tri_mask = torch.triu(torch.ones_like(tensor), diagonal=1).bool()
+
+        # Set the upper triangular part to -inf
+        tensor[upper_tri_mask] = 0
+    quant_level = calculate_quant_level(obj, iter_num)
+    if obj.training and iter_num % obj.eval_interval == 0:
+        print("quant level: ", quant_level)
+    result = tensor + quant_level * (dequantized - tensor).detach()
```
Comment (gkielian) [inline quantization/quantize.py:130]: Refactor to allow quant_level to be a floating point but also to be None, if none to go through the default path.

File: gpt_conf.py (hunk 10 above)
Comment (gkielian) [reply gpt_conf.py:7]: Add one for full_quant_iter

Comment (gkielian) [reply gpt_conf.py:7]: Then use full_quant_iter as an input to the scheduler.

**Review by gkielian — CHANGES_REQUESTED** [review]

File: gpt_conf.py [hunk 13]
```diff
@@ -207,6 +205,9 @@ class GPTConfig:
     # Quantizations
     start_quant_level: float = 0
     quant_scheduler: str = None
+    full_quant_iteration: int = 1750
```
Comment (gkielian) [inline gpt_conf.py:208]: Have this equal to none, and add assertion into code so that if quant_scheduler is set this also needs to be set

File: train.py [hunk 14]
```diff
@@ -211,7 +211,9 @@ def parse_args():
     model_group.add_argument( "--linear_std_init", type=float, default=0.02)
 
     # Quantization
-    model_group.add_argument("--start_quant_level", type=float, default=0,
+    model_group.add_argument("--full_quant_iteration", type=int, default=1750,
```
Comment (gkielian) [inline train.py:214]: Also needs to default to None

**Review by gkielian — CHANGES_REQUESTED** [review]
Looks good generally, just one requested change.

**Review by gkielian — APPROVED** [review]
